File: ruby/lib/pay_kit/operator.rb
# frozen_string_literal: true

require_relative "errors"
require_relative "signer"

module PayKit
  # Merchant identity bundle: where settled funds land (`recipient`),
  # who signs (`signer`), and whether the signer also pays the on-chain
  # network fees (`fee_payer`). Created via the `c.operator do |op| ... end`
  # block inside `PayKit.configure`, or assigned directly with
  # `c.operator = PayKit::Operator.new(...)`.
  #
  # Setters follow a deliberate "nil-as-no-op" convention so env-driven
  # configuration stays free of `if ENV[...]` guards: when the right-hand
  # side is `nil` the assignment is silently dropped and the existing
  # value (typically a default) survives. The escape hatch for actually
  # clearing a previously-set value is `op.reset!(:field)`.
  #
  # The default operator is the demo signer with `fee_payer: true` and
  # `recipient: nil`. `recipient` resolves to `signer.pubkey` via
  # `effective_recipient`, so a zero-config boot still has a settlement
  # destination. `PayKit::Config` enforces the mainnet refusal rule on
  # top of this object (see `PayKit::DemoSignerOnMainnetError`).
  class Operator
    DEFAULT_FEE_PAYER = true

    def initialize(recipient: nil, signer: nil, fee_payer: nil)
      @recipient = nil
      @signer = ::PayKit::Signer.demo
      @fee_payer = DEFAULT_FEE_PAYER

      assign_recipient(recipient)
      assign_signer(signer)
      assign_fee_payer(fee_payer)

      yield self if block_given?
    end

    attr_reader :recipient, :signer, :fee_payer

    # Nil-as-no-op setter. Non-nil values must be Strings.
    def recipient=(value)
      assign_recipient(value)
    end

    # Nil-as-no-op setter. Non-nil values must respond to the signer
    # duck-type (`#pubkey`, `#sign(message)`, `#fee_payer?`).
    def signer=(value)
      assign_signer(value)
    end

    # Nil-as-no-op setter. Non-nil values must be exactly `true` or
    # `false`; truthy coercion would mask configuration bugs.
    def fee_payer=(value)
      assign_fee_payer(value)
    end

    # `true` when the operator's signer should co-sign as Solana fee
    # payer on settlement transactions. Mirrors the boolean accessor
    # but reads predicate-style at call sites.
    def fee_payer?
      @fee_payer == true
    end

    # The address that should receive funds when a gate omits a
    # per-route `pay_to:`. Returns the explicit `recipient` when set,
    # otherwise the signer's own pubkey.
    def effective_recipient
      @recipient || @signer.pubkey
    end

    # Restore a single field to its construction default. `:recipient`
    # → nil, `:signer` → `Signer.demo`, `:fee_payer` → true. Use this
    # when the env-driven nil-no-op pattern is not enough.
    def reset!(field)
      case field
      when :recipient then @recipient = nil
      when :signer then @signer = ::PayKit::Signer.demo
      when :fee_payer then @fee_payer = DEFAULT_FEE_PAYER
      else
        raise ArgumentError, "unknown operator field #{field.inspect}; expected :recipient, :signer, or :fee_payer"
      end
      self
    end

    # Two operators are equal when their resolved recipient, signer
    # public key, and fee-payer flag all match. Used by tests and by
    # the dispatcher when it needs to detect a config change.
    def ==(other)
      other.is_a?(Operator) &&
        effective_recipient == other.effective_recipient &&
        signer.pubkey == other.signer.pubkey &&
        fee_payer? == other.fee_payer?
    end
    alias_method :eql?, :==

    def hash
      [Operator, effective_recipient, signer.pubkey, fee_payer?].hash
    end

    def to_h
      {
        recipient: effective_recipient,
        signer_pubkey: @signer.pubkey,
        signer_class: @signer.class.name,
        fee_payer: @fee_payer
      }
    end

    private

    def assign_recipient(value)
      return if value.nil?
      unless value.is_a?(String)
        raise ::PayKit::ConfigurationError, "operator.recipient must be a String, got #{value.class.name}"
      end

      @recipient = value
    end

    def assign_signer(value)
      return if value.nil?
      unless signer_like?(value)
        raise ::PayKit::ConfigurationError,
          "operator.signer must respond to #pubkey, #sign, and #fee_payer? — got #{value.class.name}"
      end

      @signer = value
    end

    def assign_fee_payer(value)
      return if value.nil?
      unless value == true || value == false
        raise ::PayKit::ConfigurationError,
          "operator.fee_payer must be true or false, got #{value.inspect}"
      end

      @fee_payer = value
    end

    def signer_like?(value)
      value.respond_to?(:pubkey) && value.respond_to?(:sign) && value.respond_to?(:fee_payer?)
    end
  end
end

File: python/src/pay_kit/operator.py
"""Merchant identity bundle: recipient, signer, and fee-payer flag."""

from __future__ import annotations

import pydantic

from pay_kit.errors import ConfigurationError
from pay_kit.signer import LocalSigner, Signer

__all__ = ["Operator"]


class Operator(pydantic.BaseModel):
    """Where settled funds land, who signs, and whether the signer pays fees.

    Mirrors the Ruby/PHP ``nil``-as-default-marker convention: a freshly
    constructed ``Operator()`` carries ``recipient=None`` / ``signer=None``,
    and :meth:`with_defaults` resolves those into the demo signer and the
    signer's own pubkey. ``recipient`` then falls back to ``signer.pubkey()``
    via :meth:`effective_recipient`, so a zero-config boot still has a
    settlement destination. ``Config`` layers the mainnet-refusal rule on top.
    """

    model_config = pydantic.ConfigDict(frozen=True, arbitrary_types_allowed=True, extra="forbid")

    recipient: str | None = None
    signer: LocalSigner | None = None
    fee_payer: bool = True

    @pydantic.field_validator("recipient", mode="before")
    @classmethod
    def _validate_recipient(cls, value: object) -> object:
        """Reject non-string recipients (``None`` stays as the default marker)."""
        if value is None:
            return None
        if not isinstance(value, str):
            raise ConfigurationError(f"operator.recipient must be a str, got {type(value).__name__}")
        return value

    @pydantic.field_validator("fee_payer", mode="before")
    @classmethod
    def _validate_fee_payer(cls, value: object) -> bool:
        """Require an exact ``bool``; truthy coercion would mask config bugs."""
        if not isinstance(value, bool):
            raise ConfigurationError(f"operator.fee_payer must be true or false, got {value!r}")
        return value

    def with_defaults(self) -> Operator:
        """Resolve ``None`` markers into shipped defaults.

        ``signer`` defaults to :meth:`Signer.demo`; ``recipient`` defaults to
        the resolved signer's pubkey. Returns a new frozen instance.
        """
        signer = self.signer if self.signer is not None else Signer.demo()
        recipient = self.recipient if self.recipient is not None else signer.pubkey()
        return Operator(recipient=recipient, signer=signer, fee_payer=self.fee_payer)

    def effective_recipient(self) -> str:
        """The settlement address: explicit ``recipient`` or the signer's pubkey."""
        if self.recipient is not None:
            return self.recipient
        signer = self.signer if self.signer is not None else Signer.demo()
        return signer.pubkey()

    def __eq__(self, other: object) -> bool:
        """Equal when resolved recipient, signer pubkey, and fee_payer all match."""
        if not isinstance(other, Operator):
            return NotImplemented
        return (
            self.effective_recipient() == other.effective_recipient()
            and self._signer_pubkey() == other._signer_pubkey()
            and self.fee_payer == other.fee_payer
        )

    def __hash__(self) -> int:
        """Hash over the resolved identity tuple (matches :meth:`__eq__`)."""
        return hash((Operator, self.effective_recipient(), self._signer_pubkey(), self.fee_payer))

    def _signer_pubkey(self) -> str:
        signer = self.signer if self.signer is not None else Signer.demo()
        return signer.pubkey()
